iml-gui/crate/src/page/filesystem.rs

```diff
@@ -1,145 +1,237 @@
 use crate::{
-    components::{alert_indicator, lock_indicator, pie_chart, table as T, Placement},
+    components::{action_dropdown, alert_indicator, lock_indicator, pie_chart, table as T, Placement},
     extensions::MergeAttrs,
     extract_id,
     generated::css_classes::C,
     route::RouteId,
-    Model, Route,
+    GMsg, Route,
 };
-use im::HashMap;
 use iml_wire_types::{
-    warp_drive::{ArcCache, ArcValuesExt},
-    Filesystem, ResourceUri, Target, TargetConfParam, TargetKind, ToCompositeId, VolumeOrResourceUri,
+    warp_drive::{ArcCache, ArcValuesExt, Locks},
+    Filesystem, Target, TargetConfParam, TargetKind, ToCompositeId, VolumeOrResourceUri,
 };
 use number_formatter as NF;
 use seed::{prelude::*, *};
-use std::sync::Arc;
-
-pub(crate) fn view<I>(model: &Model, f: &Filesystem) -> Node<I> {
-    let (mgt, mut mdt, mut ost) = model.records.target.values().filter(|t| is_fs_target(f, t)).fold(
-        (vec![], vec![], vec![]),
-        |(mut mgt, mut mdt, mut ost), t| {
-            match t.kind {
-                TargetKind::Mgt => mgt.push(t),
-                TargetKind::Mdt => mdt.push(t),
-                TargetKind::Ost => ost.push(t),
+use std::collections::{HashMap, HashSet};
+
+pub struct Row {
+    dropdown: action_dropdown::Model,
+}
+
+#[derive(Default)]
+pub struct Model {
+    pub id: u32,
+    pub mdts: Vec<Target<TargetConfParam>>,
+    pub mgt: Vec<Target<TargetConfParam>>,
+    pub osts: Vec<Target<TargetConfParam>>,
```

Comment:
Why do we need a local copy?

Reply:
The `rows` field needs to be updated whenever a target changes (add / remove). Because of this, Since we have to go through and update the rows for each table anyway, it makes sense to cache these values in the update.

We don't necessarily have to do this, but it also has the effect of being more performant. By doing it in the update, we split fields and sort only when a target has changed. If we did this in the view, the splitting and sorting would occur every render, regardless if anything relating to targets has changed.